src/stream_reader.rs

```diff
@@ -0,0 +1,172 @@
+use crate::result::IonResult;
+use crate::types::decimal::Decimal;
+use crate::types::timestamp::Timestamp;
+use crate::types::IonType;
+
+/**
+ * This trait captures the format-agnostic parser functionality needed to navigate within an Ion
+ * stream and read the values encountered into native Rust data types.
+ *
+ * Once a value has successfully been read from the stream using one of the read_* functions,
+ * calling that function again may return an Err. This is left to the discretion of the implementor.
+ */
+pub trait StreamReader {
+    /// The type returned by calls to [next], indicating the next entity in the stream.
+    /// Reader implementations representing different levels of abstraction will surface
+    /// different sets of encoding artifacts. While an application-level Reader would only surface
+    /// Ion values, a lower level Reader might surface symbol tables, Ion version markers, etc.
+    type Item;
+
+    /// The types used to represent field names, annotations, and symbol values at this Reader's
+    /// level of abstraction.
+    type Symbol;
```

Comment:
PR tour: when GATs land, we can make this generic over the lifetime of the method that constructs it. This will allow us to return (e.g.) `&str` comprised of bytes in the input buffer rather than a full `String`.